tests/helpers.py

```diff
@@ -160,3 +166,206 @@ def __init__(self, **kwargs):
         if "spec_set" in kwargs:
             self.spec_set = kwargs.pop("spec_set")
         super().__init__(spec_set=self.spec_set, **kwargs)  # type: ignore # Mixin class, this __init__ is valid
+
+
+@overload
+def isexception(obj: type[Exception] | Exception) -> Literal[True]: ...  # type: ignore[reportOverlappingOverload]
+@overload
+def isexception(obj: Any) -> Literal[False]: ...
+
+
+def isexception(obj: Any) -> bool:
+    """Check if the object is an exception."""
+    return (isinstance(obj, type) and issubclass(obj, Exception)) or isinstance(obj, Exception)
+
+
+def check_exception_equality(raised: Exception, expected: type[Exception] | Exception) -> tuple[bool, str]:
+    """Check if two exceptions are equal (compatible).
+
+    This function is used to check if the raised exception is compatible with the expected one, meaning
+    that if Excepted is a type, the raised exception must be an instance of that type. If Expected is an instance
+    of an exception then the type must be equal, but also the message and the arguments must be equal.
+
+    :param raised: The raised exception.
+    :param expected: The expected exception.
+
+    :return: Whether the exceptions are compatible.
+    """
+    if isinstance(expected, type):
+        return isinstance(
+            raised, expected
+        ), f"Did not raise the expected exception: expected {expected!r}, got {raised!r}"
+
+    if type(raised) != type(expected):
+        return check_exception_equality(raised, type(expected))  # Reuse the message from the type check
+
+    if raised.args != expected.args:
+        return False, f"Raised exception has different arguments: expected {expected.args}, got {raised.args!r}"
+
+    if str(raised) != str(expected):
+        return False, f"Raised exception has different message: expected {expected}, got {raised}"
+
+    return True, ""
```

Comment:
So, why not just use `pytest.raises`? That has a `match` kwarg which can take in a regex pattern, which is much nicer to work with than just an exact comparison.

Reply:
I liked the idea of being able to pass an initialized exception to the testgen function, and I feel like this function gives me more precision with the error message.

I can add support for regex patterns if you want.

Reply:
The issue is that sometimes, it's very annoying to have to match on the exact message, there's a reason why pytest does regex here, instead of something like that. I think we should stick to the well established `pytest.raises` than comming up with new solutions here.

Reply:
Ok, I used the initialized exception trick to pass the message to `pytest.raises`, since I think it would be less clear to pass the exception message in a tuple along with the exception type.

Reply:
hmm, but what about exceptions that aren't initialized with just a msg? (i.e. `JSONDecodeError` that needs 3 params) How will that work?

Reply:
I just pass `str(exception)` to the `match` kwarg of `pytest.raises`, how do you think it should work ?

My previous function worked for these cases but only with exact matches and I don't know what the most pytest-y way to do that would be...

Reply:
well, exceptions can override `__str__`, so that's not really good enough if we want to put some custom messages / regex patterns there. I think the func will need to accept an extra string alongside with the exceptions. This string should be optional, as we might not always want to match against some pattern (though in vast majority of cases, we should)